<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 8 RNA-Seq (an example) | Advanced R</title>
<meta name="author" content="Kálmán Abari">
<meta name="description" content="8.1 Introduction This workflow below is based on Mary Piper.1 For analysis, we used the DESeq2 package2 and also considered Michael I. Love, Simon Anders, and Wolfgang Huber3 description. The...">
<meta name="generator" content="bookdown 0.24 with bs4_book()">
<meta property="og:title" content="Chapter 8 RNA-Seq (an example) | Advanced R">
<meta property="og:type" content="book">
<meta property="og:description" content="8.1 Introduction This workflow below is based on Mary Piper.1 For analysis, we used the DESeq2 package2 and also considered Michael I. Love, Simon Anders, and Wolfgang Huber3 description. The...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 8 RNA-Seq (an example) | Advanced R">
<meta name="twitter:description" content="8.1 Introduction This workflow below is based on Mary Piper.1 For analysis, we used the DESeq2 package2 and also considered Michael I. Love, Simon Anders, and Wolfgang Huber3 description. The...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/header-attrs-2.11/header-attrs.js"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.3.0/transition.js"></script><script src="libs/bs3compat-0.3.0/tabs.js"></script><script src="libs/bs3compat-0.3.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="libs/htmlwidgets-1.5.4/htmlwidgets.js"></script><link href="libs/str_view-0.1.0/str_view.css" rel="stylesheet">
<script src="libs/str_view-binding-1.4.0/str_view.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="bs4_style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Advanced R</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> Introduction</a></li>
<li><a class="" href="warm-up-exercise.html"><span class="header-section-number">2</span> Warm-up exercise</a></li>
<li><a class="" href="rmarkdown.html"><span class="header-section-number">3</span> RMarkdown</a></li>
<li><a class="" href="advanced-data-manipulation.html"><span class="header-section-number">4</span> Advanced data manipulation</a></li>
<li><a class="" href="tidyverse-r.html"><span class="header-section-number">5</span> Tidyverse R</a></li>
<li><a class="" href="modern-graphics.html"><span class="header-section-number">6</span> Modern graphics</a></li>
<li><a class="" href="bioconductor.html"><span class="header-section-number">7</span> Bioconductor</a></li>
<li><a class="active" href="rna-seq-an-example.html"><span class="header-section-number">8</span> RNA-Seq (an example)</a></li>
<li><a class="" href="summary.html"><span class="header-section-number">9</span> Summary</a></li>
<li><a class="" href="references.html">References</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/abarik/advancedr_2021_22_1">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="rna-seq-an-example" class="section level1" number="8">
<h1>
<span class="header-section-number">8</span> RNA-Seq (an example)<a class="anchor" aria-label="anchor" href="#rna-seq-an-example"><i class="fas fa-link"></i></a>
</h1>
<div id="introduction-2" class="section level2" number="8.1">
<h2>
<span class="header-section-number">8.1</span> Introduction<a class="anchor" aria-label="anchor" href="#introduction-2"><i class="fas fa-link"></i></a>
</h2>
<p>This workflow below is based on <span class="citation">Mary Piper.<a class="footnote-ref" tabindex="0" data-toggle="popover" data-content='&lt;p&gt;&lt;span&gt;“&lt;span class="nocase"&gt;RNA-Seq with Bioconductor in R&lt;/span&gt;,”&lt;/span&gt; October 1, 2020, &lt;a href="https://learn.datacamp.com/courses/rna-seq-with-bioconductor-in-r" role="doc-biblioref"&gt;https://learn.datacamp.com/courses/rna-seq-with-bioconductor-in-r&lt;/a&gt;.&lt;/p&gt;'><sup>1</sup></a></span> For analysis, we used the <strong>DESeq2</strong> package<span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content='&lt;p&gt;Michael I. Love, Wolfgang Huber, and Simon Anders, &lt;span&gt;“&lt;span class="nocase"&gt;Moderated estimation of fold change and dispersion for &lt;span&gt;RNA&lt;/span&gt;-seq data with &lt;span&gt;DESeq&lt;/span&gt;2&lt;/span&gt;,”&lt;/span&gt; &lt;em&gt;Genome Biology&lt;/em&gt; 15, no. 12 (December 2014), &lt;a href="https://doi.org/10.1186/s13059-014-0550-8" role="doc-biblioref"&gt;https://doi.org/10.1186/s13059-014-0550-8&lt;/a&gt;.&lt;/p&gt;'><sup>2</sup></a></span> and also considered <span class="citation">Michael I. Love, Simon Anders, and Wolfgang Huber<a class="footnote-ref" tabindex="0" data-toggle="popover" data-content='&lt;p&gt;&lt;span&gt;“&lt;span class="nocase"&gt;Analyzing RNA-seq data with DESeq2&lt;/span&gt;,”&lt;/span&gt; May 12, 2020, &lt;a href="https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html" role="doc-biblioref"&gt;https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html&lt;/a&gt;.&lt;/p&gt;'><sup>3</sup></a></span> <a href="https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">description</a>.</p>
<p>The research questions and databases used in the workflow are based on the publication of <span class="citation">Casimiro Gerarduzzi et al.<a class="footnote-ref" tabindex="0" data-toggle="popover" data-content='&lt;p&gt;&lt;span&gt;“&lt;span class="nocase"&gt;Silencing &lt;span&gt;SMOC&lt;/span&gt;2 ameliorates kidney fibrosis by inhibiting fibroblast to myofibroblast transformation&lt;/span&gt;,”&lt;/span&gt; &lt;em&gt;&lt;span&gt;JCI&lt;/span&gt; Insight&lt;/em&gt; 2, no. 8 (April 2017), &lt;a href="https://doi.org/10.1172/jci.insight.90299" role="doc-biblioref"&gt;https://doi.org/10.1172/jci.insight.90299&lt;/a&gt;.&lt;/p&gt;'><sup>4</sup></a></span> RNA sequencing data can be downloaded from the Gene Expression Omnibus (GEO) database (GEO accession: [GSE85209] (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE85209" class="uri">https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE85209</a>)).</p>
</div>
<div id="loading-packages" class="section level2" number="8.2">
<h2>
<span class="header-section-number">8.2</span> Loading packages<a class="anchor" aria-label="anchor" href="#loading-packages"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb418"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load library for DESeq2</span>
<span class="co"># if (!requireNamespace("BiocManager", quietly = TRUE))</span>
<span class="co">#     install.packages("BiocManager")</span>
<span class="co"># </span>
<span class="co"># .libPaths()</span>
<span class="co"># BiocManager::install("DESeq2")</span>
<span class="co"># devtools::install_github("stephenturner/annotables")</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2">DESeq2</a></span><span class="op">)</span>

<span class="co"># Load library for RColorBrewer</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span>

<span class="co"># Load library for pheatmap</span>
<span class="co"># install.packages("pheatmap")</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span>

<span class="co"># Load library for tidyverse</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="read-counts" class="section level2" number="8.3">
<h2>
<span class="header-section-number">8.3</span> Read counts<a class="anchor" aria-label="anchor" href="#read-counts"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb419"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">smoc2_rawcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="st">"data/fibrosis_smoc2_rawcounts_unordered.csv"</span><span class="op">)</span> 
<span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span><span class="op">(</span><span class="va">smoc2_rawcounts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">smoc2_rawcounts</span><span class="op">$</span><span class="va">X</span>
<span class="va">smoc2_rawcounts</span><span class="op">$</span><span class="va">X</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>

<span class="co"># Explore the first six observations of smoc2_rawcounts</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">smoc2_rawcounts</span><span class="op">)</span>
<span class="co">#&gt;                    smoc2_fibrosis1 smoc2_fibrosis4</span>
<span class="co">#&gt; ENSMUSG00000102693               0               0</span>
<span class="co">#&gt; ENSMUSG00000064842               0               0</span>
<span class="co">#&gt; ENSMUSG00000051951              72              30</span>
<span class="co">#&gt; ENSMUSG00000102851               0               0</span>
<span class="co">#&gt; ENSMUSG00000103377               0               0</span>
<span class="co">#&gt; ENSMUSG00000104017               0               0</span>
<span class="co">#&gt;                    smoc2_normal1 smoc2_normal3</span>
<span class="co">#&gt; ENSMUSG00000102693             0             0</span>
<span class="co">#&gt; ENSMUSG00000064842             0             0</span>
<span class="co">#&gt; ENSMUSG00000051951             0             3</span>
<span class="co">#&gt; ENSMUSG00000102851             0             0</span>
<span class="co">#&gt; ENSMUSG00000103377             1             0</span>
<span class="co">#&gt; ENSMUSG00000104017             0             0</span>
<span class="co">#&gt;                    smoc2_fibrosis3 smoc2_normal4</span>
<span class="co">#&gt; ENSMUSG00000102693               0             0</span>
<span class="co">#&gt; ENSMUSG00000064842               0             0</span>
<span class="co">#&gt; ENSMUSG00000051951              36             1</span>
<span class="co">#&gt; ENSMUSG00000102851               0             0</span>
<span class="co">#&gt; ENSMUSG00000103377               0             0</span>
<span class="co">#&gt; ENSMUSG00000104017               0             0</span>
<span class="co">#&gt;                    smoc2_fibrosis2</span>
<span class="co">#&gt; ENSMUSG00000102693               0</span>
<span class="co">#&gt; ENSMUSG00000064842               0</span>
<span class="co">#&gt; ENSMUSG00000051951              51</span>
<span class="co">#&gt; ENSMUSG00000102851               0</span>
<span class="co">#&gt; ENSMUSG00000103377               0</span>
<span class="co">#&gt; ENSMUSG00000104017               0</span>

<span class="co"># Explore the structure of smoc2_rawcounts</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">smoc2_rawcounts</span><span class="op">)</span>
<span class="co">#&gt; 'data.frame':    47729 obs. of  7 variables:</span>
<span class="co">#&gt;  $ smoc2_fibrosis1: int  0 0 72 0 0 0 0 0 0 1 ...</span>
<span class="co">#&gt;  $ smoc2_fibrosis4: int  0 0 30 0 0 0 0 0 0 1 ...</span>
<span class="co">#&gt;  $ smoc2_normal1  : int  0 0 0 0 1 0 0 0 0 1 ...</span>
<span class="co">#&gt;  $ smoc2_normal3  : int  0 0 3 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;  $ smoc2_fibrosis3: int  0 0 36 0 0 0 0 0 0 1 ...</span>
<span class="co">#&gt;  $ smoc2_normal4  : int  0 0 1 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;  $ smoc2_fibrosis2: int  0 0 51 0 0 0 0 0 0 1 ...</span></code></pre></div>
</div>
<div id="distribution-of-counts" class="section level2" number="8.4">
<h2>
<span class="header-section-number">8.4</span> Distribution of counts<a class="anchor" aria-label="anchor" href="#distribution-of-counts"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb420"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">smoc2_rawcounts</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">smoc2_fibrosis1</span><span class="op">)</span>, bins <span class="op">=</span> <span class="fl">300</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="07-rnaseq_files/figure-html/unnamed-chunk-3-1.png" width="672"></div>
<div class="sourceCode" id="cb421"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">smoc2_rawcounts</span><span class="op">$</span><span class="va">smoc2_fibrosis1</span><span class="op">)</span>
<span class="co">#&gt;     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. </span>
<span class="co">#&gt;      0.0      0.0      1.0    579.7    184.0 420026.0</span></code></pre></div>
</div>
<div id="setting-of-metadata" class="section level2" number="8.5">
<h2>
<span class="header-section-number">8.5</span> Setting of metadata<a class="anchor" aria-label="anchor" href="#setting-of-metadata"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb422"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create genotype vector</span>
<span class="va">genotype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"smoc2_oe"</span>, <span class="st">"smoc2_oe"</span>, <span class="st">"smoc2_oe"</span>, <span class="st">"smoc2_oe"</span>, <span class="st">"smoc2_oe"</span>, <span class="st">"smoc2_oe"</span>, <span class="st">"smoc2_oe"</span><span class="op">)</span>

<span class="co"># Create condition vector</span>
<span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"fibrosis"</span>, <span class="st">"fibrosis"</span>, <span class="st">"fibrosis"</span>, <span class="st">"fibrosis"</span>, <span class="st">"normal"</span>, <span class="st">"normal"</span>, <span class="st">"normal"</span><span class="op">)</span>

<span class="co"># Create data frame</span>
<span class="va">smoc2_metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">genotype</span>, <span class="va">condition</span><span class="op">)</span>

<span class="co"># Assign the row names of the data frame</span>
<span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span><span class="op">(</span><span class="va">smoc2_metadata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"smoc2_fibrosis1"</span>, <span class="st">"smoc2_fibrosis2"</span>, <span class="st">"smoc2_fibrosis3"</span>, <span class="st">"smoc2_fibrosis4"</span>, <span class="st">"smoc2_normal1"</span>, <span class="st">"smoc2_normal3"</span>, <span class="st">"smoc2_normal4"</span><span class="op">)</span>

<span class="va">smoc2_metadata</span>
<span class="co">#&gt;                 genotype condition</span>
<span class="co">#&gt; smoc2_fibrosis1 smoc2_oe  fibrosis</span>
<span class="co">#&gt; smoc2_fibrosis2 smoc2_oe  fibrosis</span>
<span class="co">#&gt; smoc2_fibrosis3 smoc2_oe  fibrosis</span>
<span class="co">#&gt; smoc2_fibrosis4 smoc2_oe  fibrosis</span>
<span class="co">#&gt; smoc2_normal1   smoc2_oe    normal</span>
<span class="co">#&gt; smoc2_normal3   smoc2_oe    normal</span>
<span class="co">#&gt; smoc2_normal4   smoc2_oe    normal</span></code></pre></div>
</div>
<div id="matching-metadata-and-counts-data" class="section level2" number="8.6">
<h2>
<span class="header-section-number">8.6</span> Matching metadata and counts data<a class="anchor" aria-label="anchor" href="#matching-metadata-and-counts-data"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb423"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Use the match() function to reorder the columns of the raw counts</span>
<span class="va">reorder_idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span><span class="op">(</span><span class="va">smoc2_metadata</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">smoc2_rawcounts</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Reorder the columns of the count data</span>
<span class="va">reordered_smoc2_rawcounts</span> <span class="op">&lt;-</span> <span class="va">smoc2_rawcounts</span><span class="op">[</span> , <span class="va">reorder_idx</span><span class="op">]</span>

<span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span><span class="op">(</span><span class="va">smoc2_metadata</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">reordered_smoc2_rawcounts</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] TRUE</span>

<span class="co"># Create a DESeq2 object</span>
<span class="va">dds_smoc2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html">DESeqDataSetFromMatrix</a></span><span class="op">(</span>countData <span class="op">=</span> <span class="va">reordered_smoc2_rawcounts</span>,
                              colData <span class="op">=</span> <span class="va">smoc2_metadata</span>,
                              design <span class="op">=</span> <span class="op">~</span> <span class="va">condition</span><span class="op">)</span></code></pre></div>
</div>
<div id="normalizing-counts" class="section level2" number="8.7">
<h2>
<span class="header-section-number">8.7</span> Normalizing counts<a class="anchor" aria-label="anchor" href="#normalizing-counts"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb424"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Determine the size factors to use for normalization</span>
<span class="va">dds_smoc2</span> <span class="op">&lt;-</span> <span class="fu">estimateSizeFactors</span><span class="op">(</span><span class="va">dds_smoc2</span><span class="op">)</span>

<span class="fu">sizeFactors</span><span class="op">(</span><span class="va">dds_smoc2</span><span class="op">)</span>
<span class="co">#&gt; smoc2_fibrosis1 smoc2_fibrosis2 smoc2_fibrosis3 </span>
<span class="co">#&gt;       1.4319832       1.1189642       1.2480024 </span>
<span class="co">#&gt; smoc2_fibrosis4   smoc2_normal1   smoc2_normal3 </span>
<span class="co">#&gt;       1.0826799       0.7106482       0.7989734 </span>
<span class="co">#&gt;   smoc2_normal4 </span>
<span class="co">#&gt;       0.8482426</span>

<span class="co"># Extract the normalized counts</span>
<span class="va">smoc2_normalized_counts</span> <span class="op">&lt;-</span> <span class="fu">counts</span><span class="op">(</span><span class="va">dds_smoc2</span>, normalized<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="co"># smoc2_normalized_counts</span></code></pre></div>
</div>
<div id="quality-control---hierarchical-heatmap" class="section level2" number="8.8">
<h2>
<span class="header-section-number">8.8</span> Quality control - Hierarchical heatmap<a class="anchor" aria-label="anchor" href="#quality-control---hierarchical-heatmap"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb425"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Transform the normalized counts </span>
<span class="va">vsd_smoc2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/vst.html">vst</a></span><span class="op">(</span><span class="va">dds_smoc2</span>, blind <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># Extract the matrix of transformed counts</span>
<span class="va">vsd_mat_smoc2</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">vsd_smoc2</span><span class="op">)</span>

<span class="co"># Compute the correlation values between samples</span>
<span class="va">vsd_cor_smoc2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">vsd_mat_smoc2</span><span class="op">)</span> 

<span class="co"># Plot the heatmap</span>
<span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="va">vsd_cor_smoc2</span>, annotation <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">smoc2_metadata</span>, <span class="va">condition</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="07-rnaseq_files/figure-html/unnamed-chunk-7-1.png" width="672"></div>
</div>
<div id="quality-control---principal-component-analysis" class="section level2" number="8.9">
<h2>
<span class="header-section-number">8.9</span> Quality control - Principal component analysis<a class="anchor" aria-label="anchor" href="#quality-control---principal-component-analysis"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb426"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Transform the normalized counts </span>
<span class="va">vsd_smoc2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/vst.html">vst</a></span><span class="op">(</span><span class="va">dds_smoc2</span>, blind <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># Plot the PCA of PC1 and PC2</span>
<span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/plotPCA.html">plotPCA</a></span><span class="op">(</span><span class="va">vsd_smoc2</span>, intgroup<span class="op">=</span><span class="st">"condition"</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="07-rnaseq_files/figure-html/unnamed-chunk-8-1.png" width="672"></div>
</div>
<div id="creating-the-deseq2-object" class="section level2" number="8.10">
<h2>
<span class="header-section-number">8.10</span> Creating the DeSeq2 object<a class="anchor" aria-label="anchor" href="#creating-the-deseq2-object"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb427"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create DESeq2 object</span>
<span class="va">dds_smoc2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html">DESeqDataSetFromMatrix</a></span><span class="op">(</span>countData <span class="op">=</span> <span class="va">reordered_smoc2_rawcounts</span>,
                              colData <span class="op">=</span> <span class="va">smoc2_metadata</span>,
                              design <span class="op">=</span> <span class="op">~</span> <span class="va">condition</span><span class="op">)</span>

<span class="co"># Run the DESeq2 analysis</span>
<span class="va">dds_smoc2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html">DESeq</a></span><span class="op">(</span><span class="va">dds_smoc2</span><span class="op">)</span></code></pre></div>
</div>
<div id="deseq2-model---dispersion" class="section level2" number="8.11">
<h2>
<span class="header-section-number">8.11</span> DESeq2 model - dispersion<a class="anchor" aria-label="anchor" href="#deseq2-model---dispersion"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb428"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mean_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">smoc2_rawcounts</span>, <span class="fl">1</span>, <span class="va">mean</span><span class="op">)</span>
<span class="va">variance_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">smoc2_rawcounts</span>, <span class="fl">1</span>, <span class="va">var</span><span class="op">)</span>
<span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">mean_counts</span>, <span class="va">variance_counts</span><span class="op">)</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">mean_counts</span>, y<span class="op">=</span><span class="va">variance_counts</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"Mean counts per gene"</span>, y<span class="op">=</span><span class="st">"Vaiance per gene"</span><span class="op">)</span> </code></pre></div>
<div class="inline-figure"><img src="07-rnaseq_files/figure-html/unnamed-chunk-10-1.png" width="672"></div>
<p>Dispersion formula: <span class="math inline">\(Var=\mu+\alpha\times\mu^2\)</span></p>
<ul>
<li>
<span class="math inline">\(Var\)</span>: variance</li>
<li>
<span class="math inline">\(\mu\)</span>: mean</li>
<li>
<span class="math inline">\(\alpha\)</span>: dispersion</li>
</ul>
<div class="sourceCode" id="cb429"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Plot dispersions</span>
<span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/plotDispEsts.html">plotDispEsts</a></span><span class="op">(</span><span class="va">dds_smoc2</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="07-rnaseq_files/figure-html/unnamed-chunk-11-1.png" width="672"></div>
</div>
<div id="deseq2-model---contrasts" class="section level2" number="8.12">
<h2>
<span class="header-section-number">8.12</span> DESeq2 model - contrasts<a class="anchor" aria-label="anchor" href="#deseq2-model---contrasts"><i class="fas fa-link"></i></a>
</h2>
<p>DESeq2 Negative Binomial Model:</p>
<p><span class="math display">\[K_{ij} \sim NB(\mu_{ij},\alpha_i)\]</span>
<span class="math display">\[\mu_{ij}=s_{j}q_{ij}\]</span></p>
<p><span class="math display">\[log_2(q_{ij})=x_j\beta_j\]</span></p>
<ul>
<li>
<span class="math inline">\(K_{ij}\)</span> - raw count for gene i, samle j</li>
<li>
<span class="math inline">\(s_{ij}\)</span> - size factor</li>
<li>
<span class="math inline">\(g_{ij}\)</span> - normalized count</li>
</ul>
<div class="sourceCode" id="cb430"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Extract the results of the differential expression analysis</span>
<span class="va">smoc2_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html">results</a></span><span class="op">(</span><span class="va">dds_smoc2</span>, 
                contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"condition"</span>, <span class="st">"fibrosis"</span>, <span class="st">"normal"</span><span class="op">)</span>, 
                alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/plotMA.html">plotMA</a></span><span class="op">(</span><span class="va">smoc2_res</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="07-rnaseq_files/figure-html/unnamed-chunk-12-1.png" width="672"></div>
<div id="deseq2-rsults---lfc-shrinkage" class="section level3" number="8.12.1">
<h3>
<span class="header-section-number">8.12.1</span> DESeq2 rsults - LFC shrinkage<a class="anchor" aria-label="anchor" href="#deseq2-rsults---lfc-shrinkage"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb431"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Shrink the log2 fold change estimates to be more accurate</span>
<span class="va">smoc2_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/lfcShrink.html">lfcShrink</a></span><span class="op">(</span><span class="va">dds_smoc2</span>,
                    contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"condition"</span>, <span class="st">"fibrosis"</span>, <span class="st">"normal"</span><span class="op">)</span>,
                    res <span class="op">=</span> <span class="va">smoc2_res</span>, type <span class="op">=</span> <span class="st">"ashr"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/plotMA.html">plotMA</a></span><span class="op">(</span><span class="va">smoc2_res</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="07-rnaseq_files/figure-html/unnamed-chunk-13-1.png" width="672"></div>
</div>
</div>
<div id="deseq2-results" class="section level2" number="8.13">
<h2>
<span class="header-section-number">8.13</span> DESeq2 results<a class="anchor" aria-label="anchor" href="#deseq2-results"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb432"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">mcols</span><span class="op">(</span><span class="va">smoc2_res</span><span class="op">)</span>
<span class="co">#&gt; DataFrame with 5 rows and 2 columns</span>
<span class="co">#&gt;                        type            description</span>
<span class="co">#&gt;                 &lt;character&gt;            &lt;character&gt;</span>
<span class="co">#&gt; baseMean       intermediate mean of normalized c..</span>
<span class="co">#&gt; log2FoldChange      results log2 fold change (MM..</span>
<span class="co">#&gt; lfcSE               results posterior SD: condit..</span>
<span class="co">#&gt; pvalue              results Wald test p-value: c..</span>
<span class="co">#&gt; padj                results   BH adjusted p-values</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">smoc2_res</span><span class="op">)</span>
<span class="co">#&gt; log2 fold change (MMSE): condition fibrosis vs normal </span>
<span class="co">#&gt; Wald test p-value: condition fibrosis vs normal </span>
<span class="co">#&gt; DataFrame with 6 rows and 5 columns</span>
<span class="co">#&gt;                     baseMean log2FoldChange     lfcSE</span>
<span class="co">#&gt;                    &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;</span>
<span class="co">#&gt; ENSMUSG00000102693  0.000000       0.000000  1.317258</span>
<span class="co">#&gt; ENSMUSG00000064842  0.000000       0.000000  1.317258</span>
<span class="co">#&gt; ENSMUSG00000051951 22.478090       3.830238  0.789374</span>
<span class="co">#&gt; ENSMUSG00000102851  0.000000       0.000000  1.317258</span>
<span class="co">#&gt; ENSMUSG00000103377  0.201024      -0.144713  1.160598</span>
<span class="co">#&gt; ENSMUSG00000104017  0.000000       0.000000  1.317258</span>
<span class="co">#&gt;                         pvalue        padj</span>
<span class="co">#&gt;                      &lt;numeric&gt;   &lt;numeric&gt;</span>
<span class="co">#&gt; ENSMUSG00000102693          NA          NA</span>
<span class="co">#&gt; ENSMUSG00000064842          NA          NA</span>
<span class="co">#&gt; ENSMUSG00000051951 5.82520e-08 2.54201e-07</span>
<span class="co">#&gt; ENSMUSG00000102851          NA          NA</span>
<span class="co">#&gt; ENSMUSG00000103377 6.76672e-01          NA</span>
<span class="co">#&gt; ENSMUSG00000104017          NA          NA</span>
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">smoc2_res</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; out of 29556 with nonzero total read count</span>
<span class="co">#&gt; adjusted p-value &lt; 0.05</span>
<span class="co">#&gt; LFC &gt; 0 (up)       : 5776, 20%</span>
<span class="co">#&gt; LFC &lt; 0 (down)     : 5332, 18%</span>
<span class="co">#&gt; outliers [1]       : 15, 0.051%</span>
<span class="co">#&gt; low counts [2]     : 7207, 24%</span>
<span class="co">#&gt; (mean count &lt; 1)</span>
<span class="co">#&gt; [1] see 'cooksCutoff' argument of ?results</span>
<span class="co">#&gt; [2] see 'independentFiltering' argument of ?results</span></code></pre></div>
<div class="sourceCode" id="cb433"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Extract results</span>
<span class="va">smoc2_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html">results</a></span><span class="op">(</span><span class="va">dds_smoc2</span>, 
                contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"condition"</span>, <span class="st">"fibrosis"</span>, <span class="st">"normal"</span><span class="op">)</span>, 
                alpha <span class="op">=</span> <span class="fl">0.05</span>, 
                lfcThreshold <span class="op">=</span> <span class="fl">0.32</span><span class="op">)</span>

<span class="co"># Shrink the log2 fold changes</span>
<span class="va">smoc2_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/lfcShrink.html">lfcShrink</a></span><span class="op">(</span><span class="va">dds_smoc2</span>, 
                    contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"condition"</span>, <span class="st">"fibrosis"</span>, <span class="st">"normal"</span><span class="op">)</span>, 
                    res <span class="op">=</span> <span class="va">smoc2_res</span>, type <span class="op">=</span> <span class="st">"ashr"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb434"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Get an overview of the results    </span>
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">smoc2_res</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; out of 29556 with nonzero total read count</span>
<span class="co">#&gt; adjusted p-value &lt; 0.05</span>
<span class="co">#&gt; LFC &gt; 0 (up)       : 3716, 13%</span>
<span class="co">#&gt; LFC &lt; 0 (down)     : 3322, 11%</span>
<span class="co">#&gt; outliers [1]       : 15, 0.051%</span>
<span class="co">#&gt; low counts [2]     : 7207, 24%</span>
<span class="co">#&gt; (mean count &lt; 1)</span>
<span class="co">#&gt; [1] see 'cooksCutoff' argument of ?results</span>
<span class="co">#&gt; [2] see 'independentFiltering' argument of ?results</span></code></pre></div>
</div>
<div id="deseq2-significant-results" class="section level2" number="8.14">
<h2>
<span class="header-section-number">8.14</span> DESeq2 significant results<a class="anchor" aria-label="anchor" href="#deseq2-significant-results"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb435"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Save results as a data frame</span>
<span class="va">smoc2_res_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">smoc2_res</span><span class="op">)</span>

<span class="co"># Subset the results to only return the significant genes with p-adjusted values less than 0.05</span>
<span class="va">smoc2_res_sig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">smoc2_res_all</span>, <span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">smoc2_res_sig</span><span class="op">)</span>
<span class="co">#&gt;                      baseMean log2FoldChange      lfcSE</span>
<span class="co">#&gt; ENSMUSG00000051951   22.47809      3.8302376 0.78937355</span>
<span class="co">#&gt; ENSMUSG00000025900   12.06950     -2.0769846 0.62565698</span>
<span class="co">#&gt; ENSMUSG00000033845 1380.35712     -0.8819001 0.09788832</span>
<span class="co">#&gt; ENSMUSG00000025903 2522.97515     -1.8975565 0.15007654</span>
<span class="co">#&gt; ENSMUSG00000002459   11.55182      1.9315914 0.74833900</span>
<span class="co">#&gt; ENSMUSG00000033793 1921.19192     -0.8971164 0.08467708</span>
<span class="co">#&gt;                          pvalue         padj</span>
<span class="co">#&gt; ENSMUSG00000051951 4.698974e-07 2.829520e-06</span>
<span class="co">#&gt; ENSMUSG00000025900 5.479409e-04 2.278479e-03</span>
<span class="co">#&gt; ENSMUSG00000033845 3.838750e-09 2.815588e-08</span>
<span class="co">#&gt; ENSMUSG00000025903 1.242343e-26 2.749899e-25</span>
<span class="co">#&gt; ENSMUSG00000002459 3.154811e-03 1.160401e-02</span>
<span class="co">#&gt; ENSMUSG00000033793 3.841942e-12 3.594718e-11</span></code></pre></div>
</div>
<div id="visualization-of-results---ma-és-vulcano-plots" class="section level2" number="8.15">
<h2>
<span class="header-section-number">8.15</span> Visualization of results - MA és Vulcano plots<a class="anchor" aria-label="anchor" href="#visualization-of-results---ma-%C3%A9s-vulcano-plots"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb436"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create MA plot</span>
<span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/plotMA.html">plotMA</a></span><span class="op">(</span><span class="va">smoc2_res</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="07-rnaseq_files/figure-html/unnamed-chunk-18-1.png" width="672"></div>
<div class="sourceCode" id="cb437"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># Generate logical column </span>
<span class="va">smoc2_res_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">smoc2_res</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>threshold <span class="op">=</span> <span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span>
              
<span class="co"># Create the volcano plot</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">smoc2_res_all</span><span class="op">)</span> <span class="op">+</span> 
        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2FoldChange</span>, y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">padj</span><span class="op">)</span>, color <span class="op">=</span> <span class="va">threshold</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">"log2 fold change"</span><span class="op">)</span> <span class="op">+</span> 
        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"-log10 adjusted p-value"</span><span class="op">)</span> <span class="op">+</span> 
        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span>, 
              plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">rel</a></span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span>, hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>, 
              axis.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">rel</a></span><span class="op">(</span><span class="fl">1.25</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="07-rnaseq_files/figure-html/unnamed-chunk-18-2.png" width="672"></div>
</div>
<div id="visualization-of-results---heatmap" class="section level2" number="8.16">
<h2>
<span class="header-section-number">8.16</span> Visualization of results - heatmap<a class="anchor" aria-label="anchor" href="#visualization-of-results---heatmap"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb438"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Subset normalized counts to significant genes</span>
<span class="va">sig_norm_counts_smoc2</span> <span class="op">&lt;-</span> <span class="va">smoc2_normalized_counts</span><span class="op">[</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span><span class="op">(</span><span class="va">smoc2_res_sig</span><span class="op">)</span>, <span class="op">]</span>

<span class="co"># Choose heatmap color palette</span>
<span class="va">heat_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">6</span>, name <span class="op">=</span> <span class="st">"YlOrRd"</span><span class="op">)</span>

<span class="co"># Plot heatmap</span>
<span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="va">sig_norm_counts_smoc2</span>, 
         color <span class="op">=</span> <span class="va">heat_colors</span>, 
         cluster_rows <span class="op">=</span> <span class="cn">TRUE</span>, 
         show_rownames <span class="op">=</span> <span class="cn">FALSE</span>,
         annotation <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">smoc2_metadata</span>, <span class="va">condition</span><span class="op">)</span>, 
         scale <span class="op">=</span> <span class="st">"row"</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="07-rnaseq_files/figure-html/unnamed-chunk-19-1.png" width="672"></div>
</div>
<div id="significant-genes" class="section level2" number="8.17">
<h2>
<span class="header-section-number">8.17</span> Significant genes<a class="anchor" aria-label="anchor" href="#significant-genes"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb439"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stephenturner/annotables">annotables</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">grcm38</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 x 9</span>
<span class="co">#&gt;   ensgene  entrez symbol chr     start    end strand biotype</span>
<span class="co">#&gt;   &lt;chr&gt;     &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;   &lt;int&gt;  &lt;int&gt;  &lt;int&gt; &lt;chr&gt;  </span>
<span class="co">#&gt; 1 ENSMUSG~  14679 Gnai3  3      1.08e8 1.08e8     -1 protei~</span>
<span class="co">#&gt; 2 ENSMUSG~  54192 Pbsn   X      7.78e7 7.79e7     -1 protei~</span>
<span class="co">#&gt; 3 ENSMUSG~  12544 Cdc45  16     1.88e7 1.88e7     -1 protei~</span>
<span class="co">#&gt; 4 ENSMUSG~     NA H19    7      1.43e8 1.43e8     -1 lincRNA</span>
<span class="co">#&gt; 5 ENSMUSG~ 107815 Scml2  X      1.61e8 1.61e8      1 protei~</span>
<span class="co">#&gt; 6 ENSMUSG~  11818 Apoh   11     1.08e8 1.08e8      1 protei~</span>
<span class="co">#&gt; # ... with 1 more variable: description &lt;chr&gt;</span>

<span class="va">smoc2_res_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">smoc2_res</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span>var<span class="op">=</span><span class="st">"ensgene"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span>, y<span class="op">=</span><span class="va">grcm38</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ensgene"</span>, <span class="st">"symbol"</span>, <span class="st">"description"</span><span class="op">)</span><span class="op">]</span>,
            by<span class="op">=</span><span class="st">"ensgene"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">smoc2_res_all</span><span class="op">)</span>
<span class="co">#&gt;              ensgene   baseMean log2FoldChange     lfcSE</span>
<span class="co">#&gt; 1 ENSMUSG00000102693  0.0000000      0.0000000 1.3172581</span>
<span class="co">#&gt; 2 ENSMUSG00000064842  0.0000000      0.0000000 1.3172581</span>
<span class="co">#&gt; 3 ENSMUSG00000051951 22.4780900      3.8302376 0.7893735</span>
<span class="co">#&gt; 4 ENSMUSG00000102851  0.0000000      0.0000000 1.3172581</span>
<span class="co">#&gt; 5 ENSMUSG00000103377  0.2010237     -0.1447134 1.1605983</span>
<span class="co">#&gt; 6 ENSMUSG00000104017  0.0000000      0.0000000 1.3172581</span>
<span class="co">#&gt;         pvalue        padj        symbol</span>
<span class="co">#&gt; 1           NA          NA 4933401J01Rik</span>
<span class="co">#&gt; 2           NA          NA       Gm26206</span>
<span class="co">#&gt; 3 4.698974e-07 2.82952e-06          Xkr4</span>
<span class="co">#&gt; 4           NA          NA       Gm18956</span>
<span class="co">#&gt; 5 7.390045e-01          NA       Gm37180</span>
<span class="co">#&gt; 6           NA          NA       Gm37363</span>
<span class="co">#&gt;                                                             description</span>
<span class="co">#&gt; 1        RIKEN cDNA 4933401J01 gene [Source:MGI Symbol;Acc:MGI:1918292]</span>
<span class="co">#&gt; 2             predicted gene, 26206 [Source:MGI Symbol;Acc:MGI:5455983]</span>
<span class="co">#&gt; 3 X-linked Kx blood group related 4 [Source:MGI Symbol;Acc:MGI:3528744]</span>
<span class="co">#&gt; 4             predicted gene, 18956 [Source:MGI Symbol;Acc:MGI:5011141]</span>
<span class="co">#&gt; 5             predicted gene, 37180 [Source:MGI Symbol;Acc:MGI:5610408]</span>
<span class="co">#&gt; 6             predicted gene, 37363 [Source:MGI Symbol;Acc:MGI:5610591]</span>

<span class="va">smoc2_res_sig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">smoc2_res_all</span>, <span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span>
<span class="va">smoc2_res_sig</span> <span class="op">&lt;-</span> <span class="va">smoc2_res_sig</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">padj</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">smoc2_res_sig</span><span class="op">)</span>
<span class="co">#&gt;              ensgene  baseMean log2FoldChange      lfcSE</span>
<span class="co">#&gt; 1 ENSMUSG00000029811  8710.652       5.593885 0.11145593</span>
<span class="co">#&gt; 2 ENSMUSG00000027896  3819.931      -3.523220 0.09295523</span>
<span class="co">#&gt; 3 ENSMUSG00000062908 24293.902      -3.507301 0.11564406</span>
<span class="co">#&gt; 4 ENSMUSG00000028222  5848.681      -4.496932 0.16082443</span>
<span class="co">#&gt; 5 ENSMUSG00000032332  5162.864       4.557115 0.16437178</span>
<span class="co">#&gt; 6 ENSMUSG00000036887  3638.590       3.859276 0.13865758</span>
<span class="co">#&gt;          pvalue          padj  symbol</span>
<span class="co">#&gt; 1  0.000000e+00  0.000000e+00    Aoc1</span>
<span class="co">#&gt; 2 4.544391e-261 5.074721e-257 Slc16a4</span>
<span class="co">#&gt; 3 4.634505e-168 3.450235e-164   Acadm</span>
<span class="co">#&gt; 4 4.327404e-150 2.416206e-146   Calb1</span>
<span class="co">#&gt; 5 6.044426e-148 2.699924e-144 Col12a1</span>
<span class="co">#&gt; 6 1.030528e-144 3.835968e-141    C1qa</span>
<span class="co">#&gt;                                                                                                 description</span>
<span class="co">#&gt; 1                                    amine oxidase, copper-containing 1 [Source:MGI Symbol;Acc:MGI:1923757]</span>
<span class="co">#&gt; 2 solute carrier family 16 (monocarboxylic acid transporters), member 4 [Source:MGI Symbol;Acc:MGI:2385183]</span>
<span class="co">#&gt; 3                             acyl-Coenzyme A dehydrogenase, medium chain [Source:MGI Symbol;Acc:MGI:87867]</span>
<span class="co">#&gt; 4                                                             calbindin 1 [Source:MGI Symbol;Acc:MGI:88248]</span>
<span class="co">#&gt; 5                                             collagen, type XII, alpha 1 [Source:MGI Symbol;Acc:MGI:88448]</span>
<span class="co">#&gt; 6               complement component 1, q subcomponent, alpha polypeptide [Source:MGI Symbol;Acc:MGI:88223]</span></code></pre></div>
</div>
<div id="top-20-genes" class="section level2" number="8.18">
<h2>
<span class="header-section-number">8.18</span> Top 20 genes<a class="anchor" aria-label="anchor" href="#top-20-genes"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb440"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">top_20</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">sig_norm_counts_smoc2</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, <span class="op">]</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span>var<span class="op">=</span><span class="st">"ensgene"</span><span class="op">)</span>

<span class="va">top_20</span> <span class="op">&lt;-</span> <span class="va">top_20</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">8</span>, values_to <span class="op">=</span> <span class="st">"normalized_counts"</span>, names_to<span class="op">=</span><span class="st">"samplename"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">top_20</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 x 3</span>
<span class="co">#&gt;   ensgene            samplename      normalized_counts</span>
<span class="co">#&gt;   &lt;chr&gt;              &lt;chr&gt;                       &lt;dbl&gt;</span>
<span class="co">#&gt; 1 ENSMUSG00000051951 smoc2_fibrosis1             50.3 </span>
<span class="co">#&gt; 2 ENSMUSG00000051951 smoc2_fibrosis2             45.6 </span>
<span class="co">#&gt; 3 ENSMUSG00000051951 smoc2_fibrosis3             28.8 </span>
<span class="co">#&gt; 4 ENSMUSG00000051951 smoc2_fibrosis4             27.7 </span>
<span class="co">#&gt; 5 ENSMUSG00000051951 smoc2_normal1                0   </span>
<span class="co">#&gt; 6 ENSMUSG00000051951 smoc2_normal3                3.75</span>

<span class="va">top_20</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">inner_join</a></span><span class="op">(</span><span class="va">top_20</span>, 
                     <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span><span class="va">smoc2_metadata</span>, var<span class="op">=</span><span class="st">"samplename"</span><span class="op">)</span>,
                     by<span class="op">=</span><span class="st">"samplename"</span><span class="op">)</span>
<span class="co"># write.table(x = top_20, file = "tmp.csv", sep=";", dec=",", row.names = F, quote = T)</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">top_20</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">ensgene</span>, y<span class="op">=</span><span class="va">normalized_counts</span>, color<span class="op">=</span><span class="va">condition</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Top 20 genes"</span>, x<span class="op">=</span><span class="st">"Genes"</span>, y<span class="op">=</span><span class="st">"Normalized counts"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle<span class="op">=</span><span class="fl">45</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="07-rnaseq_files/figure-html/unnamed-chunk-21-1.png" width="672"></div>

</div>
</div>

  <div class="chapter-nav">
<div class="prev"><a href="bioconductor.html"><span class="header-section-number">7</span> Bioconductor</a></div>
<div class="next"><a href="summary.html"><span class="header-section-number">9</span> Summary</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#rna-seq-an-example"><span class="header-section-number">8</span> RNA-Seq (an example)</a></li>
<li><a class="nav-link" href="#introduction-2"><span class="header-section-number">8.1</span> Introduction</a></li>
<li><a class="nav-link" href="#loading-packages"><span class="header-section-number">8.2</span> Loading packages</a></li>
<li><a class="nav-link" href="#read-counts"><span class="header-section-number">8.3</span> Read counts</a></li>
<li><a class="nav-link" href="#distribution-of-counts"><span class="header-section-number">8.4</span> Distribution of counts</a></li>
<li><a class="nav-link" href="#setting-of-metadata"><span class="header-section-number">8.5</span> Setting of metadata</a></li>
<li><a class="nav-link" href="#matching-metadata-and-counts-data"><span class="header-section-number">8.6</span> Matching metadata and counts data</a></li>
<li><a class="nav-link" href="#normalizing-counts"><span class="header-section-number">8.7</span> Normalizing counts</a></li>
<li><a class="nav-link" href="#quality-control---hierarchical-heatmap"><span class="header-section-number">8.8</span> Quality control - Hierarchical heatmap</a></li>
<li><a class="nav-link" href="#quality-control---principal-component-analysis"><span class="header-section-number">8.9</span> Quality control - Principal component analysis</a></li>
<li><a class="nav-link" href="#creating-the-deseq2-object"><span class="header-section-number">8.10</span> Creating the DeSeq2 object</a></li>
<li><a class="nav-link" href="#deseq2-model---dispersion"><span class="header-section-number">8.11</span> DESeq2 model - dispersion</a></li>
<li>
<a class="nav-link" href="#deseq2-model---contrasts"><span class="header-section-number">8.12</span> DESeq2 model - contrasts</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#deseq2-rsults---lfc-shrinkage"><span class="header-section-number">8.12.1</span> DESeq2 rsults - LFC shrinkage</a></li></ul>
</li>
<li><a class="nav-link" href="#deseq2-results"><span class="header-section-number">8.13</span> DESeq2 results</a></li>
<li><a class="nav-link" href="#deseq2-significant-results"><span class="header-section-number">8.14</span> DESeq2 significant results</a></li>
<li><a class="nav-link" href="#visualization-of-results---ma-%C3%A9s-vulcano-plots"><span class="header-section-number">8.15</span> Visualization of results - MA és Vulcano plots</a></li>
<li><a class="nav-link" href="#visualization-of-results---heatmap"><span class="header-section-number">8.16</span> Visualization of results - heatmap</a></li>
<li><a class="nav-link" href="#significant-genes"><span class="header-section-number">8.17</span> Significant genes</a></li>
<li><a class="nav-link" href="#top-20-genes"><span class="header-section-number">8.18</span> Top 20 genes</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/abarik/advancedr_2021_22_1/blob/master/07-rnaseq.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/abarik/advancedr_2021_22_1/edit/master/07-rnaseq.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Advanced R</strong>" was written by Kálmán Abari. It was last built on 2021-10-29.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
